cogs/pug.py

```diff
@@ -606,18 +693,75 @@ def format_game_server_status(self):
+        if serverref in self.current_serverrefs() and serverref not in [None, '']:
+            self.allServers.pop(self.current_serverrefs().index(serverref))
+            return True
+        return False
+
+    def updateServerReference(self, serverref: str, serverdesc: str, serverurl: str = ''):
+        if serverref in self.current_serverrefs() and serverref not in [None, '']:
+            self.allServers.pop(self.current_serverrefs().index(serverref))
+        self.allServers.append((serverref,serverdesc,serverurl))
+        return True
+    
+    def useServer(self, index: int):
+        """Sets the active server"""
+        if index >= 0 and index <= len(self.allServers):
+            self.gameServerRef = self.allServers[index-1][0]
+            self.updateServerStatus()
+            return True
+        return False
+
     def generatePasswords(self):
         """Generates random passwords for red and blue teams."""
         # Spectator password is not changed, think keeping it fixed is fine.
         self.redPassword = RED_PASSWORD_PREFIX + str(random.randint(0, 999))
         self.bluePassword = BLUE_PASSWORD_PREFIX + str(random.randint(0, 999))
 
-    def getServerStatus(self, restrict: bool = False, delay: int = 0):
+    def getServerList(self, restrict: bool = False, delay: int = 0, allservers: bool = True):
         if restrict and (datetime.now() - self.lastUpdateTime).total_seconds() < delay:
             # 5 second delay between requests when restricted.
             return None
         
-        r = requests.post(self.postServer, headers=self.format_post_header_check)
+        body = self.format_post_body_serverref()
+
+        if allservers:
+            r = requests.post(self.postServer, headers=self.format_post_header_list)
+        else:
+            r = requests.post(self.postServer, headers=self.format_post_header_list, json=body)
+
+        self.lastUpdateTime = datetime.now()
+        if(r):
+            return r.json()
+        else:
+            return None
+
+    def validateServers(self):
+        if len(self.allServers):
+            info = self.getServerList()
+            if len(info):
+                # firstly, determine if the primary server is online and responding, then drop the local list
+                for svc in info:
+                    if svc['serverDefault'] == True and svc['serverStatus']['Summary'] not in [None,'','N/A','N/AN/A']:
+                        # Default is present and working, re-iterate through list and populate local var
+                        # If for whatever reason the default server isn't working, then stick to local list for now.
+                        self.allServers = []
+                        for sv in info:
+                            if sv['serverStatus']['Summary'] not in [None,'','N/A','N/AN/A']:
+                                self.updateServerReference(sv['serverRef'],sv['serverName'],'unreal://{0}:{1}'.format(sv['serverAddr'],sv['serverPort']))
```

Comment:
Probably not a big deal, but I avoid nested loop when possible. I'd probably do something like (indents won't format):

> serverDefaultPresent = False
> for sv in info:
>   if sv['serverDefault'] == True and sv['serverStatus']['Summary'] not in [None,'','N/A','N/AN/A']:
>     self.allServers = []
>     serverDefaultPresent = True
>     break
> for sv in info
>   if sv['serverStatus']['Summary'] not in [None,'','N/A','N/AN/A']:
>     self.updateServerReference(sv['serverRef'],sv['serverName'],'unreal://{0}:{1}'.format(sv['serverAddr'],sv['serverPort']))